runtime/tr.source/trj9/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -5381,6 +5381,257 @@ bool genInstanceOfOrCheckCastNullTest(TR::Node* node, TR::CodeGenerator* cg, TR:
+         }
+      else
+         {
+         TR::Register *offsetRegister = srm->findOrCreateScratchRegister();
+         // NOTE: In OOL helper call is not within ICF hence we can avoid passing dependency to helper call dispatch function and stretching it to merge label.
+         // Although internal control flow starts after returning from helper we need to define starting point and ending point of internal control flow.
+         cursor->setStartInternalControlFlow();
+         generateRXInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, offsetRegister, generateS390MemoryReference(dynamicCacheReg,0,cg));    
+         generateRXInstruction(cg, TR::InstOpCode::getStoreOpCode(), node, objClassReg, generateS390MemoryReference(dynamicCacheReg,offsetRegister,0,cg));
+         if (cacheCastClass)
+            generateRXInstruction(cg, TR::InstOpCode::getStoreOpCode(), node, castClassReg, generateS390MemoryReference(dynamicCacheReg,offsetRegister,addressSize,cg));
+         TR::LabelSymbol *skipResetOffsetLabel = generateLabelSymbol(cg);
+         generateRIInstruction(cg,TR::InstOpCode::getAddHalfWordImmOpCode(),node,offsetRegister,static_cast<int32_t>(cacheCastClass?addressSize*2:addressSize));
+         generateRIEInstruction(cg, TR::InstOpCode::CIJ, node, offsetRegister, snippetSizeInBytes, skipResetOffsetLabel, TR::InstOpCode::COND_BNE);
+         generateRIInstruction(cg, loadHalfWordImmOp , node, offsetRegister, addressSize);
+         generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, skipResetOffsetLabel);
+         cursor = generateRXInstruction(cg, TR::InstOpCode::getStoreOpCode(), node, offsetRegister, generateS390MemoryReference(dynamicCacheReg,0,cg));
+         TR::RegisterDependencyConditions * OOLconditions = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(0, 5, cg);
+         OOLconditions->addPostCondition(objClassReg, TR::RealRegister::AssignAny);
+         OOLconditions->addPostCondition(resultReg, TR::RealRegister::AssignAny);
+         OOLconditions->addPostCondition(dynamicCacheReg, TR::RealRegister::AssignAny);
+         OOLconditions->addPostCondition(offsetRegister, TR::RealRegister::AssignAny);
+         if (cacheCastClass)
+            OOLconditions->addPostCondition(castClassReg, TR::RealRegister::AssignAny);
+         cursor->setEndInternalControlFlow();
+         cursor->setDependencyConditions(OOLconditions);
+         srm->reclaimScratchRegister(offsetRegister);
+         }
+      srm->reclaimScratchRegister(dynamicCacheReg);
+      }
+   generateRRInstruction(cg, TR::InstOpCode::getLoadTestRegOpCode(), node, resultReg, resultReg);
+
+   // WARNING: It is not recommended to have two exit point in OOL section.
+   // In this case we need it in case of ifInstanceOf to save additional complex logic in mainline section
+   // We can take a risk of having two exit points in OOL here as there is no other register instruction between them.
```

Comment:
This seems scary to say the least. Can you please elaborate why this is safe? Register shuffles occur at the exit label of the OOL path. If you have multiple exists you may never execute the shuffles and end up with an incorrect register state.

Reply:
My understanding while implementing evaluator for instanceOf and discussion I had with you was that as there is no instruction involving registers either for data transfer or some arithmetic operation, it is unlikely that we will have a shuffle instructions in between. We had this implemented so that we can avoid unnecessary logic (Zig-Zag of branches) in main line for `ifInstanceOf`.

Reply:
> it is unlikely that we will have a shuffle instructions in between

I am unable to recall the exact conversation we had. Perhaps we can go through the logic here. I do not think _unlikely_ is a good explanation here. Either the shuffles are impossible, in which case this code is fine, or there is a scenario (however unlikely) that shuffles are generated and we emit incorrect code. As compiler developers we must ensure correctness first and foremost. This unfortunately means handling all corner cases as well, no matter how unlikely they may be.

Can we generate a log where these two exit points are seen and paste the trees here? We can discuss the correctness and whether this is safe.

Reply:
Snippet from the log file of ifInstanceOf node evaluation,
```
============================================================
; Live regs: GPR=1 FPR=0 VRF=0 GPR64=0 AR=0 {GPR_0032}
------------------------------
 n22n     (  0)  ificmpne --> block_3 BBStart at n20n ()                                              [     0x3ff9e3053e0] bci=[-1,8,53] rc=0 vc=174 vn=- li=2 udi=- nc=2 flg=0x20
 n13n     (  1)    instanceof  jitInstanceOf[#171  helper Method] [flags 0x400 0x0 ]                  [     0x3ff9e305110] bci=[-1,5,53] rc=1 vc=174 vn=- li=2 udi=- nc=2
 n12n     (  1)      aload  <parm 1 Ljava/lang/Object;>[#602  Parm] [flags 0xc0000107 0x0 ]           [     0x3ff9e3050c0] bci=[-1,4,53] rc=1 vc=174 vn=- li=2 udi=- nc=0
 n8n      (  1)      ==>aloadi (in GPR_0032) (X!=0 X>=0 isUnsignedLoad sharedMemory )
 n19n     (  1)    iconst 0 (X==0 X>=0 X<=0 )                                                         [     0x3ff9e3052f0] bci=[-1,8,53] rc=1 vc=174 vn=- li=2 udi=- nc=0 flg=0x302
------------------------------
<EVAL ificmpeq>
instanceof:Maximum Profiled Classes = 3
<EVAL aload>
[    40] O^O SPILL TEMPS: Set UpperHalfIsDead on GPR_0096
[    41] O^O SPILL TEMPS: Set UpperHalfIsDead on GPR_0112
</EVAL>
Cast Class runtimeVariable
IfInstanceOf Node : Branch True
instanceof: Class Not Evaluated. Evaluating it
instanceof: Emitting NullTest
instanceof: Loading Object Class
[    42] O^O SPILL TEMPS: Set UpperHalfIsDead on GPR_0128
instanceof: Emitting Class Equality Test
Emitting CastClassCacheTest
[    43] O^O SPILL TEMPS: Set UpperHalfIsDead on GPR_0144
instanceof: Emitting Super Class Test, Cast Class Depth = -1
[    44] O^O SPILL TEMPS: Set UpperHalfIsDead on GPR_0160
Emitting Dynamic Cache for CastClass and ObjectClass
Requested Bytes = 64
instanceof: Internal Control Flow in OOL : false
[    45] O^O SPILL TEMPS: Set UpperHalfIsDead on GPR_0176
[    46] O^O SPILL TEMPS: Set UpperHalfIsDead on GPR_0192
[    47] O^O SPILL TEMPS: Set UpperHalfIsDead on GPR_0208
[    48] O^O SPILL TEMPS: Set UpperHalfIsDead on GPR_0224
[    49] O^O SPILL TEMPS: Set UpperHalfIsDead on GPR_0240
[    50] O^O SPILL TEMPS: Set UpperHalfIsDead on GPR_0256
[    51] O^O SPILL TEMPS: Set UpperHalfIsDead on GPR_0304
[    52] O^O SPILL TEMPS: Set UpperHalfIsDead on GPR_0320
[    53] O^O SPILL TEMPS: Set UpperHalfIsDead on GPR_0336
[    54] O^O SPILL TEMPS: Set UpperHalfIsDead on GPR_0352
[    55] O^O SPILL TEMPS: Set UpperHalfIsDead on GPR_0368
[    56] O^O SPILL TEMPS: Set UpperHalfIsDead on GPR_0384
[    57] O^O SPILL TEMPS: Set UpperHalfIsDead on GPR_0400
ReturnType = Int32
[    58] O^O SPILL TEMPS: Set UpperHalfIsDead on GPR_0416
</EVAL>
------------------------------
 n22n     (  0)  ificmpne --> block_3 BBStart at n20n ()                                              [     0x3ff9e3053e0] bci=[-1,8,53] rc=0 vc=174 vn=- li=2 udi=- nc=2 flg=0x20
 n13n     (  0)    instanceof  jitInstanceOf[#171  helper Method] [flags 0x400 0x0 ]                  [     0x3ff9e305110] bci=[-1,5,53] rc=0 vc=174 vn=- li=2 udi=- nc=2
 n12n     (  0)      aload  <parm 1 Ljava/lang/Object;>[#602  Parm] [flags 0xc0000107 0x0 ] (in &GPR_0096)  [     0x3ff9e3050c0] bci=[-1,4,53] rc=0 vc=174 vn=- li=2 udi=- nc=0
 n8n      (  0)      ==>aloadi (in GPR_0032) (X!=0 X>=0 isUnsignedLoad sharedMemory )
 n19n     (  0)    iconst 0 (X==0 X>=0 X<=0 )                                                         [     0x3ff9e3052f0] bci=[-1,8,53] rc=0 vc=174 vn=- li=2 udi=- nc=0 flg=0x302
------------------------------

 [     0x3ff9e506d20]	LG      &GPR_0096, Parm[<parm 1 Ljava/lang/Object;>] ?+8(GPR5)
 [     0x3ff9e506f50]	LTGR    &GPR_0096,&GPR_0096	
 [     0x3ff9e507040]	BRC     BH(0x8), Label L0032	
 [     0x3ff9e5072b0]	LLZRGF  GPR_0128,#611 0(&GPR_0096)
 [     0x3ff9e5073a0]	CGRJ    GPR_0032,GPR_0128,Label L0017,BH(mask=0x8), 	 ; ClassEqualityTest
 [     0x3ff9e507750]	LG      GPR_0144,#612 184(GPR_0128)
 [     0x3ff9e507840]	XGR     GPR_0144,GPR_0032	
 [     0x3ff9e507930]	BRC     BH(0x8), Label L0017	
 [     0x3ff9e507a30]	CGIJ    GPR_0144,Label L0032,1,BH(mask=0x8), 	
 [     0x3ff9e507c80]	LG      GPR_0144,#613 8(GPR_0032)
 [     0x3ff9e507e40]	L       GPR_0144,#614 16(GPR_0144)
 [     0x3ff9e507f30]	NILF    GPR_0144,66048	
 [     0x3ff9e508050]	BRC     BNH(0x6), Label L0064	
 [     0x3ff9e508320]	LLGH    GPR_0160,#615 30(GPR_0032)
 [     0x3ff9e5084e0]	LLGH    GPR_0144,#616 30(GPR_0128)
 [     0x3ff9e5085d0]	CGRJ    GPR_0144,GPR_0160,Label L0032,BNL(mask=0xc), 	
 [     0x3ff9e5087c0]	LG      GPR_0144,#617 16(GPR_0128)
 [     0x3ff9e5088b0]	SLLG    GPR_0160,GPR_0160,3
 [     0x3ff9e508a80]	CG      GPR_0032,#618 0(GPR_0160,GPR_0144)
 [     0x3ff9e508b70]	BRC     BH(0x8), Label L0017	
 [     0x3ff9e508c70]	BRC     BE(0xf), Label L0032	
 [     0x3ff9e508f90]	Label L0064:	
 [     0x3ff9e509090]	LARL    GPR_0160, 0x000003FF9E508DB0	
 [     0x3ff9e5092e0]	CG      GPR_0032,#619 16(GPR_0160)
 [     0x3ff9e5093d0]	BRC     BNH(0x6), Label L0081	
 [     0x3ff9e5095a0]	LG      GPR_0144,#620 8(GPR_0160)
 [     0x3ff9e509690]	XGR     GPR_0144,GPR_0128	
 [     0x3ff9e509780]	BRC     BH(0x8), Label L0017	
 [     0x3ff9e509880]	CGIJ    GPR_0144,Label L0032,1,BH(mask=0x8), 	
 [     0x3ff9e5099a0]	Label L0081:	
 [     0x3ff9e509b70]	CG      GPR_0032,#621 32(GPR_0160)
 [     0x3ff9e509c60]	BRC     BNH(0x6), Outlined Label L0033	
 [     0x3ff9e509e30]	LG      GPR_0144,#622 24(GPR_0160)
 [     0x3ff9e509f20]	XGR     GPR_0144,GPR_0128	
 [     0x3ff9e50a010]	BRC     BH(0x8), Label L0017	
 [     0x3ff9e50a110]	CGIJ    GPR_0144,Outlined Label L0033,1,BNH(mask=0x6), 	
 [     0x3ff9e42f580]	ASSOCREGS {GPR0:D_GPR_0176:R}* {GPR1:D_GPR_0192:R}* {GPR2:GPR_0208:R} {GPR3:D_GPR_0224:R}* {GPR4:D_GPR_0240:R}* {GPR14:GPR_0256:R} {FPR0:D_FPR_0272:R}* {FPR1:D_FPR_0273:R}* {FPR2:D_FPR_0274:R}* {FPR3:D_FPR_0275:R}* {FPR4:D_FPR_0276:R}* {FPR5:D_FPR_0277:R}* {FPR6:D_FPR_0278:R}* {FPR7:D_FPR_0279:R}* {FPR8:D_FPR_0280:R}* {FPR9:D_FPR_0281:R}* {FPR10:D_FPR_0282:R}* {FPR11:D_FPR_0283:R}* {FPR12:D_FPR_0284:R}* {FPR13:D_FPR_0285:R}* {FPR14:D_FPR_0286:R}* {FPR15:D_FPR_0287:R}*
 [     0x3ff9e42f340]	Label L0032:	
 POST:
 {AssignAny:GPR_0128:R} {AssignAny:&GPR_0096:R} {AssignAny:GPR_0032:R} {AssignAny:GPR_0160:R} {AssignAny:GPR_0144:R}

------------------------------

 [     0x3ff9e50a280]	Outlined Label L0033:	
 [     0x3ff9e50d1e0]	LGR     GPR_0400,GPR13	
 [     0x3ff9e50d490]	STG     GPR5,#623 32(GPR13)
 [     0x3ff9e50d7e0]	ASSOCREGS
  PRE:
 {GPR2:GPR_0400:R} {GPR3:GPR_0032:R} {GPR4:&GPR_0096:R}
 [     0x3ff9e50d580]	BRASL   GPR_0256,0x0000000000000000	
 POST:
 [     0x3ff9e50dbc0]	ASSOCREGS {GPR2:GPR_0400:R} {GPR3:GPR_0032:R} {GPR4:&GPR_0096:R}
 [     0x3ff9e50d990]	LG      GPR5,#624 32(GPR13)
 POST:
 {GPR0:D_GPR_0176:D}* {GPR1:D_GPR_0192:D}* {GPR2:GPR_0208:D} {GPR3:D_GPR_0224:D}* {GPR4:D_GPR_0240:D}* {GPR14:GPR_0256:D} {FPR0:D_FPR_0272:D}* {FPR1:D_FPR_0273:D}*
 {FPR2:D_FPR_0274:D}* {FPR3:D_FPR_0275:D}* {FPR4:D_FPR_0276:D}* {FPR5:D_FPR_0277:D}* {FPR6:D_FPR_0278:D}* {FPR7:D_FPR_0279:D}* {FPR8:D_FPR_0280:D}* {FPR9:D_FPR_0281:D}*
 {FPR10:D_FPR_0282:D}* {FPR11:D_FPR_0283:D}* {FPR12:D_FPR_0284:D}* {FPR13:D_FPR_0285:D}* {FPR14:D_FPR_0286:D}* {FPR15:D_FPR_0287:D}* {VRF16:D_VRF_0288:D}* {VRF17:D_VRF_0289:D}*
 {VRF18:D_VRF_0290:D}* {VRF19:D_VRF_0291:D}* {VRF20:D_VRF_0292:D}* {VRF21:D_VRF_0293:D}* {VRF22:D_VRF_0294:D}* {VRF23:D_VRF_0295:D}* {VRF24:D_VRF_0296:D}* {VRF25:D_VRF_0297:D}*
 {VRF26:D_VRF_0298:D}* {VRF27:D_VRF_0299:D}* {VRF28:D_VRF_0300:D}* {VRF29:D_VRF_0301:D}* {VRF30:D_VRF_0302:D}* {VRF31:D_VRF_0303:D}* {HPR0:D_GPR_0304:D}* {HPR1:D_GPR_0320:D}*
 {HPR2:D_GPR_0336:D}* {HPR3:D_GPR_0352:D}* {HPR4:D_GPR_0368:D}* {HPR14:D_GPR_0384:D}*
 [     0x3ff9e50dd40]	LGR     GPR_0416,GPR_0208	
 [     0x3ff9e50de90]	CGIJ    GPR_0416,Label L0096,1,BH(mask=0x8), 		# (Start of internal control flow)
 [     0x3ff9e50dfb0]	OILL    GPR_0128,0x1	
 [     0x3ff9e50e0b0]	Label L0096:	
 [     0x3ff9e50e280]	LG      GPR_0144,#625 0(GPR_0160)
 [     0x3ff9e50e440]	STG     GPR_0128,#626 0(GPR_0144,GPR_0160)
 [     0x3ff9e50e600]	STG     GPR_0032,#627 8(GPR_0144,GPR_0160)
 [     0x3ff9e50e750]	AGHI    GPR_0144,0x10	
 [     0x3ff9e50e850]	CIJ     GPR_0144,Label L0097,40,BNH(mask=0x6), 	
 [     0x3ff9e50e970]	LGHI    GPR_0144,0x8	
 [     0x3ff9e50ea70]	Label L0097:	
 [     0x3ff9e50eec0]	ASSOCREGS {GPR0:D_GPR_0176:R}* {GPR1:D_GPR_0192:R}* {GPR2:GPR_0208:R} {GPR3:D_GPR_0224:R}* {GPR4:D_GPR_0240:R}* {GPR14:GPR_0256:R} {FPR0:D_FPR_0272:R}* {FPR1:D_FPR_0273:R}* {FPR2:D_FPR_0274:R}* {FPR3:D_FPR_0275:R}* {FPR4:D_FPR_0276:R}* {FPR5:D_FPR_0277:R}* {FPR6:D_FPR_0278:R}* {FPR7:D_FPR_0279:R}* {FPR8:D_FPR_0280:R}* {FPR9:D_FPR_0281:R}* {FPR10:D_FPR_0282:R}* {FPR11:D_FPR_0283:R}* {FPR12:D_FPR_0284:R}* {FPR13:D_FPR_0285:R}* {FPR14:D_FPR_0286:R}* {FPR15:D_FPR_0287:R}*
 [     0x3ff9e50ec40]	STG     GPR_0144,#628 0(GPR_0160)
 POST:
 {AssignAny:GPR_0128:R} {AssignAny:GPR_0416:R} {AssignAny:GPR_0160:R} {AssignAny:GPR_0144:R} {AssignAny:GPR_0032:R}	# (End of internal control flow)
 [     0x3ff9e42f000]	LTGR    GPR_0416,GPR_0416	
 [     0x3ff9e42f0f0]	BRC     BNH(0x6), Label L0017	
 [     0x3ff9e42f1f0]	BRC     NOP(0xf), Label L0032	

```

Reply:
If `n22n` can have GlRegDeps then they will be evaluated and stuck to the post dependency of label `L0032`. In `J9::Z::TreeEvaluator::VMifInstanceOfEvaluator` it looks like we evaluate the GlRegDeps first, then generate post dependencies and then we generate the instanceof code.

The `if` node ends a block, so from RA's perspective the register context should be empty when we start allocating this whole tree. This to me says things "should" be fine. Though there seem to be quite a lot of assumptions here. I'd feel much better if I knew we do not generated GlRegDeps when we have a fused ifinstanceof evaluation.